### ISSUE TITLE
Enhancement: [no-floating-promises] Allow await functions that are fully handled inside themselves to not have a .catch when called.

### ISSUE THREAD

**Issue by An631** [typescript-eslint/typescript-eslint#7860]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
   - [This issue ](https://github.com/typescript-eslint/typescript-eslint/issues/5543)looks related but talking about a different rule.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-floating-promises/

### Description

The purpose of this rule is to ensure no promise is **left unhandled**, in other words, every promise should have a catch statement that can handle any thrown errors as explained in the rule's documentation:

> A "floating" Promise is one that is created without any code set up to handle any errors it might throw.

This means that an async function that contains a try and catch statement (encompassing all of its content) should be considered a function that is NOT floating.

The problem right now is that if a developer wants to use the `await` - `async` syntax exclusively, this rule forces them to fall back to using `.catch()` even for functions that are completely protected inside them with a try catch statement.

My proposal is to allow functions that have a `try`-`catch` statement that covers all their content, to be called without the need of a `.catch()` statement, so these functions should not be flagged by the eslinter rule.

### Fail

```typescript
async function doAsync(): Promise<void> {
   await asyncPrint('hello world');
}

doAsync(); // Fails because the function errors are not being handled by any catch statement inside or outside.
```

### Pass

```typescript
async function doAsync(): Promise<void> {
  try {
    await asyncPrint('hello world');
  } catch(e){
    console.log('Error happened', e);
  }
}

doAsync(); // Doesn't fail because the function is fully and properly try-catch'd inside.
```

### Additional Info

The purpose of this rule is to ensure no promise is **left unhandled**, in other words, every promise should have a catch statement that can handle any thrown errors as explained in the rule's documentation:

> A "floating" Promise is one that is created without any code set up to handle any errors it might throw.

This means that an async function that contains a try and catch statement (encompassing all of its content) should be considered a function that is NOT floating.

The problem right now is that if a developer wants to use the `await` - `async` syntax exclusively, this rule forces them to fall back to using `.catch()` even for functions that are completely protected inside them with a try catch statement.

My proposal is to allow functions that have a `try`-`catch` statement that covers all of their content, to be called without the need of a `.catch()` statement

**Comment by An631** [comment]
Looking at this other issue: #5543 sounds like it is quite hard to test programmatically if a function will or can throw an error (with which I agree of course), but if we think of just checking for "catch coverage" it should be quite straightforward to verify that each line of code inside the function is wrapped in a `try`-`catch` statement which is exactly what using a '.catch()' statement at the caller level will do regardless of the internals of the function and if that function is already handling errors inside or not.

**Comment by bradzacher** [comment]
> sounds like it is quite hard to test programmatically if a function can throw or will throw an error

https://en.m.wikipedia.org/wiki/Halting_problem
It's not just quite hard - it's a famously unsolved problem within the field of computer science for almost a century...

It is possible to solve certain subsets of the problem, sure, but the general case is impossible.

To solve the problem you need to have complete information about the program and its inputs - but based on the AST + types alone you do not have complete information.

**Comment by bradzacher** [comment]
By your suggestion the following code is okay:
```ts
function doAsync() {
  try {
    await somethingAsync();
  } catch (e) {
    throw e;
  } 
}

// not a floating promise because the function body is just a try/catch
doAsync();
```

However this code will (clearly) result in an unhandled promise rejection.

What about if instead of `throw` the code is instead `return e.code.toString()`. Looks safe but what happens if `e` doesn't contain a `.code` property? Crash! Now we have seemingly safe code that again has an unhandled rejection.

Instead of us trying to encode unsafe assumptions into the lint rule by attempting to solve subsets of the halting problem, we instead leave it in your hands as the user to decide what is an acceptable risk.

Using the `void` operator to denote an intentionally un-awaited/partially handled promise is one example of how people do this. Using disable comments with an explanation is another method (and generally preferred as it documents the assumption being made for future readers).

**Comment by bradzacher** [comment]
Another reason we don't want to encode such checks into the rule is because it creates a cross-file dependency on implementation details in a function.

This is really difficult for us to build as we need to implement AST traversals outside of the regular ESLint traversal - it's also not a complete solution. For example we cannot inspect the AST across project boundaries - `.d.ts` files do not encode implementations, so the rule would fail to analyse the dependency.

This can create a frustrating experience because two usages of the same function can cause different errors depending where the call is made and whether or not TS provides us with the original ast or the declaration ast.

Its just an unstable solution to a problem on top of it being unsafe and incomplete. There are many reasons against doing this, but not many for it. As such I'm going to reject this proposal.

**Comment by An631** [comment]
For the problem of not having a throw directly in the catch section I don't think this should be considered to be a problem because the same happens right now with `.catch()` for example:

`doAsync().catch((e)=>{return e.code.toString()})`

That code is not correctly handling the errors that can happen on the catch statement and they will continue to be unhandled errors. So, my point with that is that we don't necessarily have to verify for any errors inside of the catch statements because that is the same behavior we already have in the current solution with .catch().

The second problem you mentioned of creating cross-file dependencies I totally understand and see why it can be a difficult thing to process with ASTs. 

So, to summarize, I wouldn't consider this proposal an incomplete or unsafe solution because it would have the same functionality we currently have with `.catch()` which is basically a try-catch that wraps around the whole function, but it is hard to implement given the cross-file checks problem and most likely not a good approach as you mention.

I appreciate you taking the time to review the proposal, thanks!
